File: examples/noiseloop-exportmp4/noiseloop-exportmp4.go
package main

import (
	"context"
	"strconv"
	"time"

	"github.com/emprcl/runal"
)

const (
	duration = 5
	radius   = 4
)

var seed1, seed2 int64

func main() {
	seed1 = time.Now().Unix()
	seed2 = seed1 + 1000
	runal.Run(context.Background(), setup, draw, runal.WithOnKey(onKey))
}

func setup(c *runal.Canvas) {
	c.CellModeDouble()
}

func draw(c *runal.Canvas) {
	c.Clear()

	theta := c.LoopAngle(duration)
	c.NoiseSeed(seed1)
	noise := c.NoiseLoop(theta, 1)
	c.NoiseSeed(seed2)
	noise2 := c.NoiseLoop(theta, 1)
	x := c.Map(noise, 0, 1, radius, float64(c.Width-radius))
	y := c.Map(noise2, 0, 1, radius, float64(c.Height-radius))
	color := c.Map(noise, 0, 1, 0, 10)
	colorStr := strconv.FormatFloat(color, 'f', -1, 64)
	colorBg := c.Map(noise2, 0, 1, 200, 210)
	colorBgStr := strconv.FormatFloat(colorBg, 'f', -1, 64)

	c.Background("#", colorBgStr, colorBgStr)
	c.Stroke(" ", colorStr, colorStr)
	c.Fill(" ", colorStr, colorStr)
	c.Circle(int(x), int(y), radius)
}

func onKey(c *runal.Canvas, e runal.KeyEvent) {
	if e.Key == "space" {
		seed1 = time.Now().Unix()
		seed2 = seed1 + 1000
	}
	if e.Key == "c" {
		c.SaveCanvasToMP4("flash.mp4", duration)
	}
}

File: examples/noiseloop-exportmp4/noiseloop-exportmp4.js
let duration = 5;
let radius = 4;
let seed1 = Date.now();
let seed2 = seed1 + 1000;

function setup(c) {
  c.cellModeDouble();
}

function draw(c) {
  c.clear();

  let theta = c.loopAngle(duration);
  c.noiseSeed(seed1);
  let noise = c.noiseLoop(theta, 1);
  c.noiseSeed(seed2);
  let noise2 = c.noiseLoop(theta, 1);
  let x = c.map(noise, 0, 1, radius, c.width - radius);
  let y = c.map(noise2, 0, 1, radius, c.height - radius);
  let color = c.map(noise, 0, 1, 0, 10);
  let colorBg = c.map(noise2, 0, 1, 200, 210);

  c.background("#", colorBg, colorBg);
  c.stroke(" ", color, color);
  c.fill(" ", color, color);
  c.circle(x, y, radius);
}

function onKey(c, e) {
  if (e.key == "space") {
    seed1 = Date.now();
    seed2 = seed1 + 1000;
  }
  if (e.key == "c") {
    c.saveCanvasToMP4("flash.mp4", duration);
  }
}
